Cabal/Distribution/Simple/Doctest.hs

```diff
@@ -62,36 +72,102 @@ doctest pkg_descr lbi suffixes doctestFlags = do
      --   smsg = setupMessage' verbosity "Running Doctest on" (packageId pkg_descr)
      --          (componentLocalName clbi) (maybeComponentInstantiatedWith clbi)
 
-     case component of 
+     case component of
        CLib lib -> do
-         args <- DoctestArgs . map snd <$> getLibSourceFiles verbosity lbi lib clbi
-         runDoctest verbosity doctestProg args
+         withTempDirectoryEx verbosity tmpFileOpts (buildDir lbi) "tmp" $
+           \tmp -> do
+             inFiles <- map snd <$> getLibSourceFiles verbosity lbi lib clbi
+             args    <- mkDoctestArgs verbosity tmp lbi clbi inFiles (libBuildInfo lib)
+             runDoctest verbosity (compiler lbi) (hostPlatform lbi) doctestProg args
        CExe exe -> do
-         args <- DoctestArgs . map snd <$> getExeSourceFiles verbosity lbi exe clbi
-         runDoctest verbosity doctestProg args
+         withTempDirectoryEx verbosity tmpFileOpts (buildDir lbi) "tmp" $
+           \tmp -> do
+             inFiles <- map snd <$> getExeSourceFiles verbosity lbi exe clbi
+             args    <- mkDoctestArgs verbosity tmp lbi clbi inFiles (buildInfo exe)
+             runDoctest verbosity (compiler lbi) (hostPlatform lbi) doctestProg args
        CFLib _  -> return () -- do not doctest foreign libs
        CTest _  -> return () -- do not doctest tests
-       CBench _ -> return () -- do not doctest benchmarks 
-     
+       CBench _ -> return () -- do not doctest benchmarks
+--
+--
+componentGhcOptions :: Verbosity -> LocalBuildInfo
+                 -> BuildInfo -> ComponentLocalBuildInfo -> FilePath
+                 -> GhcOptions
+componentGhcOptions verbosity lbi bi clbi odir =
+  let f = case compilerFlavor (compiler lbi) of
+            GHC   -> GHC.componentGhcOptions
+            GHCJS -> GHCJS.componentGhcOptions
+            _     -> error $
+                       "Distribution.Simple.Doctest.componentGhcOptions:" ++
+                       "doctest only supports GHC and GHCJS"
+  in f verbosity lbi bi clbi odir
+
+mkDoctestArgs :: Verbosity
+              -> FilePath
+              -> LocalBuildInfo
+              -> ComponentLocalBuildInfo
+              -> [FilePath]
+              -> BuildInfo
+              -> IO DoctestArgs
+mkDoctestArgs verbosity tmp lbi clbi inFiles bi = do
+  let vanillaOpts = (componentGhcOptions normal lbi bi clbi (buildDir lbi))
```

Comment:
Since there is modest duplication here with existing code (which is not easy to remove, so don't) I'd like forward and backward pointers between the code, as a reminder that if someone changes one you might have to change the other.